### PR TITLE
[AMDGPU] Make max dwords of memory cluster configurable

### PR DESCRIPTION
**Pull request by ruiling**
We find it helpful to increase the value for graphics workload. Make it configurable so we can experiment with a different value. It might be more helpful we can have a per-function value, but I am not sure how this can be done properly.

### CODE REVIEW

**Review by arsenm — APPROVED** [review]
function attribute

**Comment by ruiling** [comment]
> function attribute

Thanks! What's the suggested way to access that value in function attribute from the SIInstrInfo? Maybe keep a record of that value in MachineFunctionInfo and get from it?

**Comment by arsenm** [comment]
> > function attribute
> 
> Thanks! What's the suggested way to access that value in function attribute from the SIInstrInfo? Maybe keep a record of that value in MachineFunctionInfo and get from it?

That's just a matter of caching it in MFI or whether read directly from the IR attribute. In either case, the MachineFunction needs to be passed into the hook

**Comment by ruiling** [comment]
> > > function attribute
> > 
> > 
> > Thanks! What's the suggested way to access that value in function attribute from the SIInstrInfo? Maybe keep a record of that value in MachineFunctionInfo and get from it?
> 
> That's just a matter of caching it in MFI or whether read directly from the IR attribute. In either case, the MachineFunction needs to be passed into the hook

Added a function attribute for it, do we want to keep the command line option for it?

File: llvm/lib/Target/AMDGPU/SIMachineFunctionInfo.cpp [hunk 1]
```diff
@@ -163,6 +163,11 @@ SIMachineFunctionInfo::SIMachineFunctionInfo(const Function &F,
   if (!S.empty())
     S.consumeInteger(0, HighBitsOf32BitAddress);
 
+  A = F.getFnAttribute("amdgpu-max-memory-cluster-dwords");
+  S = A.getValueAsString();
+  if (!S.empty())
+    S.consumeInteger(0, MaxMemoryClusterDWords);
```
Comment (arsenm) [inline llvm/lib/Target/AMDGPU/SIMachineFunctionInfo.cpp:169]: Use getFnAttributeAsParsedInteger

File: llvm/lib/Target/AMDGPU/SIInstrInfo.cpp [hunk 2]
```diff
@@ -559,33 +554,39 @@ bool SIInstrInfo::shouldClusterMemOps(ArrayRef<const MachineOperand *> BaseOps1,
...
   if (!BaseOps1.empty() && !BaseOps2.empty()) {
     const MachineInstr &FirstLdSt = *BaseOps1.front()->getParent();
     const MachineInstr &SecondLdSt = *BaseOps2.front()->getParent();
     if (!memOpsHaveSameBasePtr(FirstLdSt, BaseOps1, SecondLdSt, BaseOps2))
       return false;
+
+    const SIMachineFunctionInfo *MFI =
+        FirstLdSt.getMF()->getInfo<SIMachineFunctionInfo>();
+    if (MFI->getMaxMemoryClusterDWords())
+      MaxMemoryClusterDWords = MFI->getMaxMemoryClusterDWords();
```
Comment (arsenm) [inline llvm/lib/Target/AMDGPU/SIInstrInfo.cpp:567]: Should just directly take MFI->getMaxMemoryClusterDWords()

**Review by arsenm — COMMENTED** [review]
None of the mir test changes parse the value?

**Comment by ruiling** [comment]
> None of the mir test changes parse the value?

Do you mean parsing the value from function attribute? I will add one.

File: llvm/lib/Target/AMDGPU/SIInstrInfo.cpp (hunk 2 above)
Comment (ruiling) [reply llvm/lib/Target/AMDGPU/SIInstrInfo.cpp]: I have fixed this by adding a default limit value in SIInstrInfo.h, which could be accessible from both places.

**Comment by ruiling** [comment]
> > None of the mir test changes parse the value?
> 
> Do you mean parsing the value from function attribute? I will add one.
I just added a small test to parse the value from MFI.

**Review by arsenm — APPROVED** [review]

**Review by arsenm — APPROVED** [review]

**Pull request merged**